packages/core/database/lib/entity-manager/relations-orderer.js

```diff
@@ -1,6 +1,105 @@
+  // Map to validate if relation is already in sortedConnect or DB.
+  const relInArray = initialArr.reduce((acc, rel) => ({ ...acc, [rel.id]: true }), {});
+  // Map to store the first index where a relation id is connected
+  const firstSeen = {};
+  // Map to validate if connect relation has already been computed
+  const computedIdx = {};
+
+  connectArr.forEach((rel, idx) => {
+    // If adjacent relation is not in the database or seen yet in the connect array
+    // then we need to sort the connect array
+    const adjacentRelId = rel.position?.before || rel.position?.after;
+    if (!relInArray[adjacentRelId] && !firstSeen[adjacentRelId]) needsSorting = true;
+    // Populate firstSeen
+    if (!(rel.id in firstSeen)) firstSeen[rel.id] = idx;
+  });
+
+  // If we don't need to sort the connect array, we can return it as is
+  if (!needsSorting) return connectArr;
+
+  // Iterate over connectArr and populate sortedConnect
+
+  connectArr.forEach((rel, idx) => {
+    const pushRelation = (rel) => {
+      sortedConnect.push(rel);
+      relInArray[rel.id] = true;
+    };
+
+    const computeRelation = (rel, relationsSeenInBranch) => {
+      const adjacentRelId = rel.position?.before || rel.position?.after;
+
+      // This connect has already been computed
+      if (idx in computedIdx) return;
+
+      if (!adjacentRelId || relInArray[adjacentRelId]) {
+        return pushRelation(rel);
+      }
+
+      // If the relation has already been seen in the current branch,
+      // it means there is a circular reference
+      if (relationsSeenInBranch[adjacentRelId]) {
+        throw new InvalidRelationError(
+          'A circular reference was found in the connect array. ' +
+            'One relation is trying to connect before/after another one that is trying to connect before/after it'
+        );
+      }
+
+      // Look if id is referenced elsewhere in the array
+      const adjacentRelIdx = firstSeen[adjacentRelId];
+
+      if (adjacentRelIdx) {
+        const adjacentRel = connectArr[adjacentRelIdx];
+
+        // Mark adjacent relation idx as computed,
+        // so it is not computed again later in the loop
+        computedIdx[adjacentRelIdx] = true;
+        computeRelation(adjacentRel, { ...relationsSeenInBranch, [rel.id]: true });
+        pushRelation(rel);
+      } else if (strictSort) {
+        // If we reach this point, it means that the adjacent relation is not in the connect array
+        // and it is not in the database. This should not happen.
+        throw new InvalidRelationError(
+          `There was a problem connecting relation with id ${rel.id} at position ${JSON.stringify(
+            rel.position
+          )}. The relation with id ${adjacentRelId} needs to be connected first.`
+        );
+      } else {
+        // We are in non-strict mode so we can push the relation.
+        pushRelation({ id: rel.id, position: { end: true } });
+      }
+    };
+
+    computeRelation(rel, {});
+  });
```

Comment:
Might it be worth defining these functions outside the forEach so they are only created once? I think you'd need to pass the `idx` to `computeRelation` like this
```suggestion
  const pushRelation = (rel) => {
    sortedConnect.push(rel);
    relInArray[rel.id] = true;
  };

  const computeRelation = (rel, idx, relationsSeenInBranch) => {
    const adjacentRelId = rel.position?.before || rel.position?.after;

    // This connect has already been computed
    if (idx in computedIdx) return;

    if (!adjacentRelId || relInArray[adjacentRelId]) {
      return pushRelation(rel);
    }

    // If the relation has already been seen in the current branch,
    // it means there is a circular reference
    if (relationsSeenInBranch[adjacentRelId]) {
      throw new InvalidRelationError(
        'A circular reference was found in the connect array. ' +
          'One relation is trying to connect before/after another one that is trying to connect before/after it'
      );
    }

    // Look if id is referenced elsewhere in the array
    const adjacentRelIdx = firstSeen[adjacentRelId];

    if (adjacentRelIdx) {
      const adjacentRel = connectArr[adjacentRelIdx];

      // Mark adjacent relation idx as computed,
      // so it is not computed again later in the loop
      computedIdx[adjacentRelIdx] = true;
      computeRelation(adjacentRel, idx, { ...relationsSeenInBranch, [rel.id]: true });
      pushRelation(rel);
    } else if (strictSort) {
      // If we reach this point, it means that the adjacent relation is not in the connect array
      // and it is not in the database. This should not happen.
      throw new InvalidRelationError(
        `There was a problem connecting relation with id ${rel.id} at position ${JSON.stringify(
          rel.position
        )}. The relation with id ${adjacentRelId} needs to be connected first.`
      );
    } else {
      // We are in non-strict mode so we can push the relation.
      pushRelation({ id: rel.id, position: { end: true } });
    }
  };

  connectArr.forEach((rel, idx) => {
    computeRelation(rel, idx, {});
  });
```

Reply:
good point, made the change!